File: algovisual/frontend/src/selection.py
import pygame
import random
import time

pygame.init()
WIDTH = 600
HEIGHT = 400
BAR_WIDTH = 80
FPS = 10
screen = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.set_caption("Selection Sort Visualizer")

WHITE = (255, 255, 255)
RED = (255, 0, 0)
GREEN = (0, 255, 0)
BLUE = (0, 0, 255)
BACKGROUND_COLOR = WHITE  

def draw_bars(arr, highlight=None):
    screen.fill(BACKGROUND_COLOR)  
    for i, val in enumerate(arr):
        color = BLUE if i != highlight else RED
        pygame.draw.rect(screen, color, (i * (BAR_WIDTH + 10), HEIGHT - val * 50, BAR_WIDTH, val * 50))
    pygame.display.update()
def selection_sort(arr):
    for i in range(len(arr)):
        min_idx = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
            draw_bars(arr, highlight=j)  
            pygame.time.delay(1000)  
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
        draw_bars(arr, highlight=i) 
        pygame.time.delay(1000) 


def main():
    arr = random.sample(range(1, 9), 5)
    
    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
        draw_bars(arr)
        pygame.time.delay(1000)

        selection_sort(arr)

        time.sleep(2)
        running = False

    pygame.quit()

if __name__ == "__main__":
    main()

File: algovisual/frontend/src/selection.js
let values = [];
let i = 0, j = 0, minIndex;

function setup() {
  createCanvas(600, 400);
  values = Array.from({ length: 5 }, () => random(height));
  minIndex = i;
  frameRate(2); 
}
function draw() {
  background(220);
  for (let k = 0; k < values.length; k++) {
    if (k === i) fill(255, 0, 0); 
    else if (k === minIndex) fill(0, 255, 0); 
    else fill(100);
    rect(k * (width / values.length), height - values[k], width / values.length - 2, values[k]);
  }

  if (i < values.length) {
    if (j < values.length) {
      if (values[j] < values[minIndex]) {
        minIndex = j; 
      }
      j++;
    } else {
      [values[i], values[minIndex]] = [values[minIndex], values[i]];
      i++;
      j = i;
      minIndex = i;
    }
  } else {
    noLoop(); 
  }
}